### PR TITLE
Refactor ProgressTracker and other stateless methods into separate utils files, and add unit tests to handle empty log_metrics() case.

### PR DESCRIPTION
**Pull request by justinxzhao**

### CODE REVIEW

File: ludwig/utils/trainer_utils.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            "steps": self.steps,
+            "last_improvement_steps": self.last_improvement_steps,
+            "learning_rate": self.learning_rate,
+            "best_valid_metric": self.best_eval_metric,
+            "num_reductions_lr": self.num_reductions_learning_rate,
+            "num_increases_bs": self.num_increases_batch_size,
+        }
+        for metrics_dict_name in [
+            "train_metrics",
+            "vali_metrics",
```
Comment (brightsparc) [inline ludwig/utils/trainer_utils.py:130]: More common to use `val_metrics` here or use the follow word`validation_metrics` ?

File: tests/ludwig/utils/test_trainer_utils.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    assert progress_tracker.log_metrics() == {
+        "batch_size": 5,
+        "best_valid_metric": 0,
+        "epoch": 0,
+        "last_improvement_steps": 0,
+        "learning_rate": 0.01,
+        "num_increases_bs": 0,
+        "num_reductions_lr": 0,
+        "steps": 0,
+        "vali_metrics.combined.loss": 0.2,
```
Comment (brightsparc) [inline tests/ludwig/utils/test_trainer_utils.py:62]: As above.

File: ludwig/utils/trainer_utils.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        best_eval_metric=best_eval_metric,
+        best_reduce_learning_rate_eval_metric=best_reduce_learning_rate_eval_metric,
+        last_reduce_learning_rate_eval_metric_improvement=0,
+        best_increase_batch_size_eval_metric=best_increase_batch_size_eval_metric,
+        last_increase_batch_size_eval_metric_improvement=0,
+        learning_rate=learning_rate,
+        num_reductions_learning_rate=0,
+        num_increases_batch_size=0,
+        train_metrics=initialize_trainer_metric_dict(output_features),
+        vali_metrics=initialize_trainer_metric_dict(output_features),
```
Comment (brightsparc) [inline ludwig/utils/trainer_utils.py:48]: As below

**Review by brightsparc — APPROVED** [review]
LGTM. Minor bit about using `vali` abbreviation for validation as being non-standard

File: ludwig/utils/trainer_utils.py (hunk 1 above)
Comment (justinxzhao) [reply ludwig/utils/trainer_utils.py]: Thanks for the suggestion. Changed to `validation_metrics` everywhere.
